web/docs/internal/gestures/gesture-input-serialization.md

```diff
@@ -0,0 +1,27 @@
+# Gesture Input Serialization
+
+The GestureRecognizer engine is designed to handle and process asynchronous, continuous input of one or more touchpoints at the same time.  The processing of events takes time, though, and on older devices it is possible for new input events to come in while older events are still being processed.  As it is possible for an earlier event to affect how a later event should be processed, it is necessary to ensure that all events are processed in a specific order. We wish to capture the user's original intent for each aspect of input and later match it against the system state that a user would reasonably anticipate for the input based on their previous interactions, even if such a state hasn't yet actually been reached or activated at the true time of input.
+
+For a keyboard, suppose a user taps the `Shift` and `A` keys with minimal time gap between the two events.  The `Shift` state may not be in place yet once the `A` is tapped - event processing does take _some_ time, after all.  The serialization scheme used by the engine will delay processing for the `A` until the `Shift` state _has_ taken effect, leading a capital `A` to be emitted instead of a lowercase `a`.
```

Comment:
I am confused by this. The gesture engine operates in the web browser's single-threaded model, because we aren't using workers for this. Events come in sequentially, in the order that they are received. The time events take to be processed should be irrelevant in terms of the model (ignoring performance concerns here) -- because processing is single threaded. Earlier events are guaranteed to be received by the Javascript code before later events.

I just don't understand how the problem described here can arise -- unless we are introducing deferral with promises -- but even then, if we consistently defer every incoming event with a promise, then the order would still be automatically enforced.

Reply:
> unless we are introducing deferral with promises

We are.  And keep in mind, I'm literally describing the old TWo caps problem that was a _very_ real issue for a long time.
- #7173

Javascript touch events note the exact element at the time of a touch input - even if that input is no longer in the DOM by the time the event is actually raised and processed.  In this case, both Shift and A are default layer - JS events _will_ still get the lowercase layer's A without specialized handling.

Reply:
OK, but that's _not_ what you describe in the document. You are describing event serialization, not dealing with stale DOM data from events.